codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java

```diff
@@ -157,31 +167,71 @@ public ChannelFuture handshake(Channel channel) {
      *
      * @param channel
      *            Channel
+     * @param handlers
+     *            the HTTP specific {@link ChannelHandler}s to remove. This must contain either at least
+     *            {@link HttpClientCodec} or {@link HttpResponseDecoder} and {@link HttpRequestEncoder}. Other handlers
+     *            like {@link HttpObjectAggregator} and {@link HttpContentDecompressor} may be present as well.
+     */
+    public ChannelFuture handshake(Channel channel, Set<? extends ChannelHandler> handlers) {
+        return handshake(channel, handlers, channel.newPromise());
+    }
+
+    /**
+     * Begins the opening handshake
+     *
+     * @param channel
+     *            Channel
      * @param promise
      *            the {@link ChannelPromise} to be notified when the opening handshake is sent
+     * @deprecated use {@link #handshake(Channel, Set, ChannelPromise)}
      */
+    @Deprecated
     public final ChannelFuture handshake(Channel channel, final ChannelPromise promise) {
-        FullHttpRequest request =  newHandshakeRequest();
-
-        HttpResponseDecoder decoder = channel.pipeline().get(HttpResponseDecoder.class);
-        if (decoder == null) {
-            HttpClientCodec codec = channel.pipeline().get(HttpClientCodec.class);
-            if (codec == null) {
-               promise.setFailure(new IllegalStateException("ChannelPipeline does not contain " +
-                       "a HttpResponseDecoder or HttpClientCodec"));
-               return promise;
-            }
+        Set<ChannelHandler> handlers = new HashSet<ChannelHandler>();
+        addIfNotNull(handlers, channel.pipeline(), HttpResponseDecoder.class);
+        addIfNotNull(handlers, channel.pipeline(), HttpRequestEncoder.class);
+        addIfNotNull(handlers, channel.pipeline(), HttpClientCodec.class);
+        addIfNotNull(handlers, channel.pipeline(), HttpObjectAggregator.class);
+        addIfNotNull(handlers, channel.pipeline(), HttpContentDecompressor.class);
+
+        return handshake(channel, handlers, promise);
+    }
+
+    /**
+     * Begins the opening handshake
+     *
+     * @param channel
+     *            Channel
+     * @param handlers
+     *            the HTTP specific {@link ChannelHandler}s to remove. This must contain either at least
+     *            {@link HttpClientCodec} or {@link HttpResponseDecoder} and {@link HttpRequestEncoder}. Other handlers
+     *            like {@link HttpObjectAggregator} and {@link HttpContentDecompressor} may be present as well.
+     * @param promise
+     *            the {@link ChannelPromise} to be notified when the opening handshake is sent
+     * @deprecated use {@link #handshake(Channel, Set, ChannelPromise)}
+     */
+    public final ChannelFuture handshake(Channel channel, final Set<? extends ChannelHandler> handlers,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the missing @Deprecated annotation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)